### ISSUE TITLE
FaceLivenessDetectorView blank view

### ISSUE THREAD

**Issue by devedup** [aws-amplify/amplify-ui-swift-liveness#30]
### Describe the bug

On presenting the FaceLivenessDetectorView view, it's just a blank screen. No error logs, just logs outputting that it has successfully connected to Cognito

### Steps To Reproduce

```swift
Just add the view and present it
```

### Expected behavior

see something

### Amplify Framework Version

latest

### Amplify Categories

API

### Dependency manager

Swift PM

### Swift version

5

### CLI version

latest

### Xcode version

latest

### Relevant log output

```shell
<details>
<summary>Log Messages</summary>


INSERT LOG MESSAGES HERE
```
</details>
```


### Is this a regression?

No

### Regression additional context

_No response_

### Device

iphone 13 pro max

### iOS Version

iOS 14

### Specific to simulators

_No response_

### Additional context

_No response_

**Comment by devedup** [comment]
oops.. better delete that AWS key now...

**Comment by atierian** [comment]
> oops.. better delete that AWS key now... 😅

I also deleted the revision history, so it's no longer visible. 🙂

> My app already has authentication separately to this, and I just need to be able to do face liveness checks

If you're using Amplify Auth, the auth portion is handled for you. If you're not, you have to inject the credentials provider. Based on your example, your doing the right thing here.

> I ran amplify update auth again and chose somethings other than the default settings, and I'm now seeing the 'Liveness Check' instructions. 👍

If you're not using Amplify, you don't need to do this. Since you're passing in a credentials provider, the liveness component doesn't interact with Amplify Auth.

> However, on tapping 'Being Check', as soon as it sees my face, the web socket drops:
`FaceLivenessDetectionError(code: 7, message: "Websocket connection unexpectedly closed", recoverySuggestion: "")`

What you're seeing here is a fallback error that you generally shouldn't see. We recently made some improvements around in this area that reports more informative and actionable errors, but it hasn't been released yet. The branch [fix/error-handling](https://github.com/aws-amplify/amplify-ui-swift-liveness/tree/fix/error-handling) has those changes. I'm going to make a quick change there and let you know when I'm done. If you target that branch (temporarily until this is released) you should get a more descriptive error message.

In the meantime, there are two things I recommend checking:
1. Does the role associated with the credentials you're using have the necessary IAM permissions? (see [Update IAM Role Permissions](https://ui.docs.amplify.aws/swift/connected-components/liveness#update-iam-role-permissions)). 
2. Is the Session ID still valid? Session IDs are valid for only one use and have a very short lifetime (I believe it's 2 minutes currently) after creating them with the [CreateFaceLivenessSession](https://docs.aws.amazon.com/rekognition/latest/APIReference/API_CreateFaceLivenessSession.html) API.

**Comment by devedup** [comment]
Thanks for deleting the revision history. I've deleted the associated user now anyway, as it was a temp user I created for testing. 

Looking forward to those changes - I'll try out that branch. 

1. Yes, I temporarily created an IAM roll that gives admin access to everything, just to get this working.. and then I'll create a more appropriate role later. 
2. Yes - I have a REST client running connected to my back end, so i generate a new session ID regularly. 

Thanks for your time with this.

**Comment by atierian** [comment]
@devedup following up on your question back in the issue you opened.

https://github.com/aws-amplify/amplify-ui-swift-liveness/issues/28#issuecomment-1551964565

> Thanks. Yes, currently I have an AWS IAM user with this inline policy ``` { "Version": "2012-10-17", "Statement": [ { "Effect": "Allow", "Action": "rekognition:StartFaceLivenessSession", "Resource": "*" }, { "Effect": "Allow", "Action": "rekognition:CreateFaceLivenessSession", "Resource": "*" } ] } ``` … but it still would mean embedding the key/secret in my code - or returning them from a backend call after auth….. which still exposes them. The use case is that I have users that are already authenticated on a dating app, and they can opt in to have their profile/account photos verified. I’m already using Recognition to detect faces match successfully, but need to add another layer to stop fake photos getting through. Hence using Face Liveness to detect real people.

Let's take a closer look at this.

> but it still would mean embedding the key/secret in my code

You don't need to embed them in the app, nor do we recommend doing so.

> The use case is that I have users that are already authenticated on a dating app

If you have users authenticated in your app, you are already returning some form of credentials to your app to authorize any requests that user makes, no? Are those generated by AWS? If so, you can attach a policy for the Liveness APIs to the applicable role that your users assume. In that case, the credentials you provide to the Liveness component are the same that your user already has on device to authorize requests to your other endpoints.

**Comment by atierian** [comment]
> Got it, thanks for the info.
> 
> You can target the branch `fix/error-handling` now. That should give you some more insights into what exactly is going wrong. Let me know if you have any questions.

Also wanting to follow up on this. Have you tried this branch? What error are you seeing?

**Comment by devedup** [comment]
Hi, Yes I did start using that branch immediately and I was receiving the error 12. I believe it was down to the session token being invalid. Once I fixed that, I've been able to test successfully now. 


So my authentication is not via AWS, it's separate mobile number and OTP code, which then gives the user a short lived JWT to authorise requests. That being said, the FaceLivenessView needs AWS credentials to function - so I do need to pass some back somehow.

I currently pass the AWS credentials back in an authenticated call...,, I could go an extra step and obfuscate them somehow, maybe encrypt them with the JWT Session Token... although not sure what I'm really gaining by doing this. 

The AWS credential only has one permission (shown below), which is 'StartFacelivenessSession'. Now a bad actor/spammer, or whoever I'm trying to protect from getting access to this has a few options. They can find the credentials returned from the auth call and then all they can do with it is spam the StartFacelivenessSession endpoint - for what purpose or goal? I need to think like a bad actor 😂 If they were linked to an AWS/Cognito user policy... it's the same difference isn't it?? They create an account, and then have access to spam the StartFacelivenessSession. 

Is there something I'm missing with best practice as to why it's so important to hide these? If I'm returning them in the auth call, I could change them at any point, which blocks any abuse, similarly how I could block access to an AWS user if there was abuse. 

Please clarify if I'm missing anything obvious here. 🤔

`{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "rekognition:StartFaceLivenessSession",
            "Resource": "*"
        }
    ]
}`

**Comment by atierian** [comment]
> Once I fixed that, I've been able to test successfully now.

Awesome, happy to hear you got it working.

> So my authentication is not via AWS, it's separate mobile number and OTP code, which then gives the user a short lived JWT to authorise requests.

Understood. From what you've shared about of your session management and your concerns, it looks like there's a pretty straightforward approach you can take here.

You can have an endpoint that checks for authorization based on the user's JWT (the one you manage) and does the following:
- Invokes [CreateFaceLivenessSession](https://docs.aws.amazon.com/rekognition/latest/APIReference/API_CreateFaceLivenessSession.html)
- Fetches temporary credentials. Since you're currently not using AWS credentials / session management, the easiest way will likely be [STS](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html).
- Returns temporary credentials and the session ID.

> They can find the credentials returned from the auth call and then all they can do with it is spam the StartFacelivenessSession endpoint.

Not really, no. A `StartFaceLivenessSession` request requires a valid session ID (in addition to authorization). A request that doesn't contain a valid session ID will fail. Session IDs have a very short lifetime; I believe it's 2 minutes. Additionally, they can only be used once before becoming invalid. If you want to do any additional checking, this will need to be done on your side before you give a valid session ID to a user. Does that alleviate your concerns?

**Comment by devedup** [comment]
STS... that's a new one for me. There always seems to be an AWS service for what I require. Brilliant, thanks.

**Comment by devedup** [comment]
So I'm working through STS, it's been a bit of a pain getting IAM roles and policies setup, but I'm at the stage where it's returning me temporary credentials. However these credentials don't seem to work to (rekognition:StartFaceLivenessSession) , which I'm working through with AWS support. 

However, it should be of note to you that when I use credentials that don't work, the app just crashes and doesn't fail gracefully with an error code. I've done a pull on error-handling branch, but there isn't anything new that I'm missing.

**Comment by atierian** [comment]
Thanks for following up. Can you send a full crash report? It should be handling that situation gracefully and just completing with an error if the credentials are invalid. The crash report will help us identify what's going wrong to make a fix.

Thanks!

**Comment by devedup** [comment]
I'll have a look. One thing I just noticed this morning is that I need to continue with the error branch you asked me to use.

If I go back to main branch, it enforces me to add an Auth plugin before configuration...i.e. 

`        // AWS Amplify
        do {
            Amplify.Logging.logLevel = .verbose
           // try Amplify.add(plugin: AWSCognitoAuthPlugin())
            try Amplify.configure()
        } catch {
            print("An error occurred setting up Amplify: \(error)")
        }`

Which makes no sense, as I don't use Cognito. It will crash on starting the detector if I don't do this, but also when I do do this because I don't have the cognito config setup.

**Comment by atierian** [comment]
Just want to check in to see if you've resolved the crash. If not, can you please send the requested crash report? Thanks!

The improved error messaging was included in version [1.0.1](https://github.com/aws-amplify/amplify-ui-swift-liveness/releases/tag/1.0.1).

You shouldn't need to use use Amplify Auth if you're injecting your own credentials provider. I wasn't able to reproduce this though. We recommend depending on the component use versions instead of targeting `main`. As `main` can contain changes that aren't finalized before releasing.

**Comment by atierian** [comment]
I see that you provided the crash report in another issue. Thanks. 
I'm closing this one now as the original issue is resolved. We'll follow up on the new issue you opened.
